spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -982,15 +982,49 @@ struct ProductionVisitor
     // Generate code to synchronize on the given production. We assume that the
     // given production supports some form of lookahead; if the production is
     // not supported an error will be generated.
-    void syncProduction(const Production& p) {
+    void syncProduction(const Production& p_) {
+        const auto* p = &p_;
+
+        if ( auto resolved = p->tryAs<production::Resolved>() ) {
+            p = &grammar.resolved(*resolved);
+        }
+
         // Validation.
-        if ( auto while_ = p.tryAs<production::While>(); while_ && while_->expression() )
+        if ( auto while_ = p->tryAs<production::While>(); while_ && while_->expression() )
             hilti::logger().error("&synchronize cannot be used on while loops with conditions");
 
-        auto tokens = grammar.lookAheadsForProduction(p);
+        // Handle synchronization via `synchronize-at` or `synchronize-after` unit properties.
+        if ( const auto& unit = p->tryAs<production::Unit>() ) {
+            const auto& type = unit->unitType();
+            const auto synchronize_at = type.propertyItem("%synchronize-at");
+            const auto synchronize_after = type.propertyItem("%synchronize-after");
+
+            std::optional<Expression> e;
+
+            if ( synchronize_at )
+                e = synchronize_at->expression();
+
+            if ( synchronize_after )
+                e = synchronize_after->expression();
```

Comment:
This will behave in unexpected ways if both `%synchronize-at` and `%synchronize-after` are given in the same unit. We should reject such setups.